src/NuGet.Clients/PackageManagement.VisualStudio/SolutionRestore/SolutionRestoreWorker.cs

```diff
@@ -146,163 +134,171 @@ public void CleanCache()
             SolutionRestoreRequest request,
+            bool blockingUi,
             CancellationToken token)
         {
-            // Start the restore job in a separate task on a background thread
-            // it will switch into main thread when necessary.
-            var joinableTask = Task.Run(
-                () => StartRestoreJobAsync(request, restoreOperation.BlockingUI, token));
+            _activeLock.Wait(token);
 
-            await joinableTask
-                .ContinueWith(t => restoreOperation.ContinuationAction(t));
-
-            return await restoreOperation;
-        }
-
-        private async Task PromoteTaskToActiveAsync(BackgroundRestoreOperation restoreOperation, CancellationToken token)
-        {
-            var pendingTask = (Task<bool>)restoreOperation;
-
-            int attempt = 0;
-            for (var retry = true;
-                retry && !token.IsCancellationRequested && attempt != SanePromoteAttemptsLimit;
-                attempt++)
+            BackgroundRestoreOperation newRestoreOperation;
+            try
             {
-                // Grab local copy of active task
-                var activeTask = _activeRestoreTask;
+                var activeRestoreOperation = _activeRestoreOperation;
 
-                // Await for the completion of the active *unbound* task
-                var cancelTcs = new TaskCompletionSource<bool>();
-                using (var ctr = token.Register(() => cancelTcs.TrySetCanceled()))
-                {
-                    await Task.WhenAny(activeTask, cancelTcs.Task);
-                }
+                newRestoreOperation = BackgroundRestoreOperation.Start(
+                    () => StartRestoreJobAsync(request, blockingUi, activeRestoreOperation, token));
 
-                // Try replacing active task with the new one.
-                // Retry from the beginning if the active task has changed.
-                retry = Interlocked.CompareExchange(
-                    ref _activeRestoreTask, pendingTask, activeTask) != activeTask;
+                _activeRestoreOperation = newRestoreOperation;
             }
-
-            if (attempt == SanePromoteAttemptsLimit)
+            finally
             {
-                throw new InvalidOperationException("Failed promoting pending task.");
+                _activeLock.Release();
             }
+
+            return newRestoreOperation;
         }
 
         private async Task<bool> StartRestoreJobAsync(
-            SolutionRestoreRequest jobArgs, bool blockingUi, CancellationToken token)
+            SolutionRestoreRequest request,
+            bool blockingUi,
+            BackgroundRestoreOperation activeRestoreOperation,
+            CancellationToken token)
         {
+            await TaskScheduler.Default;
+
+            await activeRestoreOperation;
+
+            token.ThrowIfCancellationRequested();
+
+            DrainPendingRequestQueue(token);
```

Comment:
This seems reasonable if all restores are solution wide, but does this also need to combine the settings on each request? For example if one request has force=true in the queue shouldn't the next restore force?